crates/apollo-parser/src/ast/node_ext.rs

```diff
@@ -193,12 +196,164 @@ fn unescape_string(input: &str) -> String {
     output
 }
 
+const ESCAPED_TRIPLE_QUOTE: &str = r#"\""""#;
+const TRIPLE_QUOTE: &str = r#"""""#;
+
+fn is_block_string(input: &str) -> bool {
+    input.starts_with(TRIPLE_QUOTE)
+}
+
+/// Iterator over the lines in a GraphQL string, using GraphQL's definition of newlines
+/// (\r\n, \n, or just \r).
+struct GraphQLLines<'a> {
+    input: &'a str,
+    finished: bool,
+}
+
+impl<'a> GraphQLLines<'a> {
+    fn new(input: &'a str) -> Self {
+        Self {
+            input,
+            finished: false,
+        }
+    }
+}
+
+impl<'a> Iterator for GraphQLLines<'a> {
+    type Item = &'a str;
+    fn next(&mut self) -> Option<Self::Item> {
+        // Can't just check for the input string being empty, as an empty string should still
+        // produce one line.
+        if self.finished {
+            return None;
+        }
+
+        let Some(index) = memchr::memchr2(b'\r', b'\n', self.input.as_bytes()) else {
+            self.finished = true;
+            return Some(self.input);
+        };
+        let line = &self.input[..index];
+        let rest = match self.input.get(index..=index + 1) {
+            Some("\r\n") => &self.input[index + 2..],
+            _ => &self.input[index + 1..],
+        };
+        self.input = rest;
```

Comment:
The above is fine. It can be marginally nicer with `split_at` and `strip_prefix`:
```suggestion
        let (line, rest) = self.input.split_at(index);
        // `rest` starts with \r or \n
        self.input = rest.strip_prefix("\r\n").unwrap_or_else(|| &rest[1..])
```

Reply:
my personal feeling is that `split_at` looks nicer but the manual `match` is more obviously correct despite being a bit noisy... though as i look back and forth i don't feel strongly anymore.